File: programmers/cpp/programmers_heap_2.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int stock, vector<int> dates, vector<int> supplies, int k) {
	int answer = 0;
	int idx = 0;
	priority_queue<pair<int, int>> q;

	for (int i = 0; i < k; i++)
	{
		if (i == dates[idx])
		{
			q.push(make_pair(supplies[idx], dates[idx]));
			idx++;
		}
		if (stock == 0)
		{
			stock += q.top().first;
			q.pop();
			answer++;
		}
		stock--;
	}

	return answer;
}

File: programmers/python/programmers_heap_2.py
import heapq as hq

def solution(stock, dates, supplies, k):
	answer = 0
	idx = 0
	pq = []
	while True:
		if stock >= k:
			break
		for i in range(idx, len(dates)):
			if dates[i] <= stock:
				hq.heappush(pq, -supplies[idx])
				idx += 1
			else:
				break
		stock -= (hq.heappop(pq))
		answer += 1
	return answer